File: armour/src/main/java/com/yangzhiwen/armour/ArmourIContentProvider.java
package com.yangzhiwen.armour;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.util.Arrays;

/**
 * use java
 * Created by yangzhiwen on 2017/8/19.
 */

public class ArmourIContentProvider implements InvocationHandler {

    private ArmourHacker armourHacker;
    /**
     * IContentProvider
     */
    private Object icp;

    public ArmourIContentProvider(Object icp, ArmourHacker armourHacker) {
        this.icp = icp;
        this.armourHacker = armourHacker;
    }

    @Override
    public Object invoke(Object o, Method method, Object[] objects) throws Throwable {
        System.out.println("ArmourIContentProvider invoke " + method.getName() + Arrays.toString(objects));

        armourHacker.onIContentProviderInvoke(objects);


        return method.invoke(icp, objects);
    }
}

File: armour/src/main/java/com/yangzhiwen/armour/ArmourIContentProvider.kt
//package com.yangzhiwen.armour
//
//import java.lang.reflect.InvocationHandler
//import java.lang.reflect.InvocationTargetException
//import java.lang.reflect.Method
//
///**
// * Created by yangzhiwen on 2017/8/19.
// */
//class ArmourIContentProvider(val icp: Any?) : InvocationHandler {
//    var count = 0
//
//    override fun invoke(proxy: Any?, method: Method, args: Array<out Any>): Any? {
//        println("ArmourIContentProvider invoke - $method - $args")
//
//
//        try {
//            return method.invoke(icp, args)
//        } catch (e: InvocationTargetException) {
//            e.printStackTrace()
//            throw e.getTargetException()
//        }
//
//
////        if (args == null) return method.invoke(icp)
////        else return method.invoke(icp, args)
//    }
//}